contracts/Governance/Timelock.sol

```diff
@@ -1,123 +1,174 @@
-        require(getBlockTimestamp() <= eta + GRACE_PERIOD, "Timelock::executeTransaction: Transaction is stale.");
+    function executeTransactionsBatch(Transaction[] memory transactions, uint256 eta) external payable onlyExecutorOrOwner {
+        _executeTransactionsBatchInternal(transactions, eta);
+    }
 
-        queuedTransactions[txHash] = false;
+    function approveAndExecuteTransactionsBatch(Transaction[] memory transactions, uint256 eta) external payable onlyOwner {
+        bytes32 txParamsHash = keccak256(abi.encode(transactions, eta));
 
-        bytes memory callData;
+        _approveTransactionsBatchInternal(txParamsHash);
+        _executeTransactionsBatchInternal(transactions, eta);
+    }
 
-        if (bytes(signature).length == 0) {
-            callData = data;
-        } else {
-            callData = abi.encodePacked(bytes4(keccak256(bytes(signature))), data);
-        }
+    function approveAndExecuteTransactionsBatchRaw(bytes calldata txParamsData) external payable onlyOwner {
+        bytes32 txParamsHash = keccak256(txParamsData);
+        (Transaction[] memory transactions, uint256 eta) = abi.decode(txParamsData, (Transaction[], uint256));
 
-        // Execute the call
-        (bool success, bytes memory returnData) = target.call{value: value}(callData);
-        require(success, "Timelock::executeTransaction: Transaction execution reverted.");
+        _approveTransactionsBatchInternal(txParamsHash);
+        _executeTransactionsBatchInternal(transactions, eta);
+    }
+
+    function _approveTransactionsBatchInternal(bytes32 txParamsHash) internal onlyOwner {
+        require(queuedTransactions[txParamsHash] == TransactionStatus.Queued, "Timelock: transaction is not queued");
+
+        queuedTransactions[txParamsHash] = TransactionStatus.Approved;
+
+        emit ApprovedTransactionsBatch(txParamsHash);
+    }
+
+    function _executeTransactionsBatchInternal(Transaction[] memory transactions, uint256 eta) internal {
+        bytes32 txParamsHash = keccak256(abi.encode(transactions, eta));
 
-        emit ExecuteTransaction(txHash, target, value, signature, data, eta);
+        require(queuedTransactions[txParamsHash] == TransactionStatus.Approved, "Timelock: Transaction hasn't been approved.");
+        require(block.timestamp >= eta, "Timelock: Transaction hasn't surpassed the execution delay.");
+        require(block.timestamp <= eta + executionGracePeriod, "Timelock: Transaction is stale.");
+
+        delete queuedTransactions[txParamsHash];
+
+        for (uint256 i = 0; i < transactions.length; i++) {
+            (
+                bool success, /* ignore the rest */
+
+            ) = transactions[i].recipient.call{value: transactions[i].value}(transactions[i].data);
+            require(success, "Timelock: Transaction execution reverted.");
+
+            emit ExecutedTransaction(txParamsHash, transactions[i].recipient, transactions[i].value, transactions[i].data, eta);
```

Comment:
A sidenote: If we don't want it, we would also have to unify min and max delays.

ETA has 2 consequences:

1. We can give different delays to different changes (small delay for minor changes, big eta for breaking changes) - we could be more fair to the users giving them more time to leave the protocol in case of braking changes
2. We can schedule multiple transactions with the same parameters (distinguished by ETA)

It doesn't really complicate the calls - we have helper functions which extract all the parameters for us. TxHash (of the queue transaction) is all we need.

It's not critical though. We can remove it if you want.

Reply:
My bad here, too much time in-front of the code I missed the part we use it in our `requires` :)